### PR TITLE
Enable flake8

### PR DESCRIPTION
**Pull request by hmaarrfk**
Flake8 was enabled, but we were ignoring the output so it might as well have been removed.
That said, it is useful to check unused variables, and undefined variables.

I disabled most of the style options that don't make sense when you write mathematical code,
added a few `# noqa` to keep the useful warnings for future code.

Things are mostly split by logical modifications in a few different commits 

I setup flake8 such that it would be much less strict than before and only run on a single build 

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by hmaarrfk** [comment]
@scikit-image/core I don't know if you care about flake8, but I find it gives much better error message than running code through python when you have a syntax error. The commit message are now much improved compared to the first version I submitted a few months ago.

File: .travis.yml [hunk 1]
```diff
@@ -38,7 +38,7 @@ matrix:
   include:
     - os: linux
       python: 3.5
-      env: OPTIONAL_DEPS=1 WITH_PYSIDE=1 BUILD_DOCS=1
+      env: OPTIONAL_DEPS=1 WITH_PYSIDE=1 BUILD_DOCS=1 TEST_FLAKE8=1
```
Comment (soupault) [inline .travis.yml]: `RUN_FLAKE8` or smth?

File: .travis.yml [hunk 2]
```diff
@@ -79,6 +79,15 @@ before_install:
     - which python; python --version
     - tools/build_versions.py
     - tools/check_sdist.py
+    # It is really annoying if flake8 causes your tests to fail first
+    # but in the event that you have a typo, flake8 is pretty good at
+    # catching it.
+    # Therefore, it is best to run flake8 first, on a single build
+    # with the oldest version of python we support
+    - if [[ $TEST_FLAKE8 == "1" ]]; then
```
Comment (soupault) [inline .travis.yml]: Should all of this be in a dedicated `tools\flake8.py`?

File: .travis.yml [hunk 3]
```diff
@@ -79,6 +79,15 @@ before_install:
...
+        pip install flake8;
+        flake8;
```
Comment (soupault) [inline .travis.yml]: You probably need to keep the same flake8 arguments as before.

File: skimage/io/_plugins/util.py [hunk 4]
```diff
@@ -6,11 +6,8 @@
 
 # utilities to make life easier for plugin writers.
 
-try:
-    import multiprocessing
-    CPU_COUNT = multiprocessing.cpu_count()
-except:
-    CPU_COUNT = 2
+import multiprocessing
```
Comment (soupault) [inline skimage/io/_plugins/util.py]: Already in. https://github.com/scikit-image/scikit-image/pull/3434

File: .travis.yml (hunk 3 above)
Comment (hmaarrfk) [reply .travis.yml]: I think it is all in the `setup.cfg`. I added a dummy commit to trip up flake8 to test it.

File: .travis.yml (hunk 2 above)
Comment (hmaarrfk) [reply .travis.yml]: done. `flake8.sh`

File: .travis.yml (hunk 1 above)
Comment (hmaarrfk) [reply .travis.yml]: ok.

**Comment by hmaarrfk** [comment]
@soupault see this build https://travis-ci.org/scikit-image/scikit-image/jobs/436291887#L905 showing we don't need to setup any flake8 command line arguments.

**Comment by hmaarrfk** [comment]
Yeah. Let me think about this again. Flake8 is a little bit of an annoyance when it causes something to fail for no reason. This is especially true when they change their mind about the style guide and whatnot. 

The worst part about this script, is that if somebody else introduces a flake8 error, all future contributors would see red. 

My goal is to have contributors get to green as easily as possible. The goal of these tests should be to guide contributors through the mundane stuff (style, and test infrastructure), while helping them with the technical parts.

**Comment by soupault** [comment]
@hmaarrfk well, a build failing with a redundant whitespace is not what I would like to see either. But having flake8 executed during the early phase of CI routine is pretty useful.
You're right, flake8 is PR-agnostic in this case, and may lead to a chain of red builds.

What is we adjust pep8speaks bot to perform only the most significant checks (as, currently, specified in the PR), and remove flake8 completely (or, at least, disable its influence on the build status)?

Other changes in the PR are very welcome.

File: setup.cfg [hunk 5]
```diff
@@ -1,2 +1,47 @@
 [metadata]
 license_file = LICENSE.txt
+
+[flake8]
+ignore =
```
Comment (Borda) [inline setup.cfg]: you can use full names instead of shortcuts

File: setup.cfg [hunk 6]
```diff
@@ -1,2 +1,47 @@
...
+    F841,
+    # F401 'package' imported but unused. This shows up too much in __init__ files
+    F401,
+    # F403 'from .manage_plugins import *' used; unable to detect undefined names
+    F403,
+    # W391 blank line at end of file
+    W391
+
+exclude =
+    test_*,
```
Comment (Borda) [inline setup.cfg]: why excluding tests?

File: skimage/io/__init__.py [hunk 7]
```diff
@@ -4,7 +4,9 @@
 
 """
 
-from .manage_plugins import *
+from .manage_plugins import (use_plugin, call_plugin, plugin_info,
+                             plugin_order, reset_plugins,
+                             find_available_plugins, available_plugins)
 from .sift import *
```
Comment (Borda) [inline skimage/io/__init__.py]: why here is not the names list too?

File: skimage/measure/mc_meta/createluts.py [hunk 8]
```diff
@@ -134,10 +134,11 @@ def search_line(lines, refline, start=0):
 
 
 
-def getLutNames(prefix):
+def getLutNames(prefix, luts):
```
Comment (Borda) [inline skimage/measure/mc_meta/createluts.py]: from the review scope, it is not clear but `luts` does not seems to be used

File: skimage/segmentation/random_walker_segmentation.py [hunk 9]
```diff
@@ -30,7 +30,7 @@ def new_del(self):
             pass
     umfpack.UmfpackContext.__del__ = new_del
     UmfpackContext = umfpack.UmfpackContext()
-except:
+except:  # noqa
```
Comment (Borda) [inline skimage/segmentation/random_walker_segmentation.py]: an exception should be specified et leas by `Exception`

File: setup.cfg (hunk 5 above)
Comment (hmaarrfk) [reply setup.cfg]: ??? How?

Comment (Borda) [reply setup.cfg]: my bad, it seems that it works only for pylint

File: setup.cfg (hunk 6 above)
Comment (hmaarrfk) [reply setup.cfg]: Tests often include many data arrays, that are well justified not to follow pep8.

File: skimage/io/__init__.py (hunk 7 above)
Comment (hmaarrfk) [reply skimage/io/__init__.py]: you could. I think i gave up on changing style, I mostly like flake8 to find python typos, and not imposing style.

File: setup.cfg (hunk 6 above)
Comment (jni) [reply setup.cfg]: @hmaarrfk this is true, but test functions themselves should follow PEP8. I'd prefer finer-grained control here. Maybe it's not doable but I prefer to ignore these warnings rather than silence them completely.

File: skimage/segmentation/random_walker_segmentation.py (hunk 9 above)
Comment (jni) [reply skimage/segmentation/random_walker_segmentation.py]: I agree, and here we can probably do even better. @JDWarner might have some thoughts?

File: setup.cfg (hunk 6 above)
Comment (hmaarrfk) [reply setup.cfg]: I really believe that warnings should be set to 0, except when they would add more noise (like in examples).

The idea is that currently the flake8 stuff is just fully being ignore.

File: skimage/segmentation/random_walker_segmentation.py (hunk 9 above)
Comment (hmaarrfk) [reply skimage/segmentation/random_walker_segmentation.py]: These kind of maintenance changes really require 3 devs working according to our policies. Will not work with such large sweeping changes without focused effort ;)

**Comment by stefanv** [comment]
Love this; thanks @hmaarrfk! Needs a quick rebase, hopefully not too honerous (err, may well be).

Reminds me about the conversations around black, @jni.  I'd be okay with the way it's used in Napari, e.g.

**Comment by jni** [comment]
Yeah, pre-commit hooks are the right way to implement these things. Unfortunately, though, it makes it harder to get started developing for skimage...

**Comment by hmaarrfk** [comment]
THe flake8 changes I had to actually enact are things like tabulation, and trailing whitespace, I see these as "minor" and better to deal with near the end of the PR, instead of upfront with a pre-commit hook

**Comment by stefanv** [comment]
> THe flake8 changes I had to actually enact are things like tabulation, and trailing whitespace, I see these as "minor" and better to deal with near the end of the PR, instead of upfront with a pre-commit hook

Agreed, for now. My suggestion was for the long term, where I think we would benefit from enforced compliance that simultaneously relieves the burden on the contributor.

I don't think it's much of a problem to help new users with this.  For other projects, I have a single command they can run to install the commit hook, e.g. `make lint-install` (or, perhaps, `python .tools/install-lint.py`.

File: doc/release/contribs.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python
```
Comment (grlee77) [inline doc/release/contribs.py:1]: Did this file get added to the PR by accident? It does not seem related to anything else here.

**Comment by grlee77** [comment]
This now needs a rebase, but I would be fine with merging once the CI is green. We can always refine things again later as needed.

**Comment by stefanv** [comment]
Implemented using ruff in #6729
